tests/karate/src/test/java/common/net/MessageQueue.java

```diff
@@ -0,0 +1,124 @@
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class MessageQueue implements MessageBuffer {
+
+  private final List<String> queue = new LinkedList<>();
+
+  public synchronized void onNewMsg(String msg) {
+    queue.add(msg);
+    notifyAll();
+  }
+
+  @Override
+  public synchronized String peek() {
+    return queue.get(0);
+  }
+
+  @Override
+  public synchronized String peek(final Predicate<String> filter) {
+    return queue.stream().filter(filter).findFirst().orElse(null);
+  }
+
+  @Override
+  public synchronized List<String> peekAll() {
+    return new ArrayList<>(queue);
+  }
+
+  @Override
+  public synchronized List<String> peekAll(Predicate<String> filter) {
+    return queue.stream().filter(filter).collect(Collectors.toList());
+  }
+
+  @Override
+  public synchronized List<String> peekN(int firstN) {
+    return queue.stream().limit(firstN).collect(Collectors.toList());
+  }
+
+  @Override
+  public synchronized String take() {
+    return queue.isEmpty() ? null : queue.remove(0);
+  }
+
+  @Override
+  public synchronized String take(Predicate<String> filter) {
+    Iterator<String> it = queue.iterator();
+    while (it.hasNext()) {
+      String msg = it.next();
+
+      if (filter.test(msg)) {
+        it.remove();
+        return msg;
+      }
+    }
+
+    return null;
+  }
+
+  @Override
+  public synchronized List<String> takeAll() {
+    List<String> messages =  new ArrayList<>(queue);
+    queue.clear();
+    return messages;
+  }
+
+  @Override
+  public synchronized List<String> takeAll(final Predicate<String> filter) {
+    List<String> messages = new LinkedList<>();
+    Iterator<String> it = queue.iterator();
+
+    while (it.hasNext()) {
+      String msg = it.next();
+      if (filter.test(msg)) {
+        messages.add(msg);
+        it.remove();
+      }
+    }
+
+    return messages;
+  }
+
+  @Override
+  public synchronized List<String> takeN(int limit) {
+    List<String> messages = new LinkedList<>();
+    for (int i = 0; i < limit && !queue.isEmpty(); i++)
+      messages.add(queue.remove(0));
+
+    return messages;
+  }
+
+  @Override
+  public synchronized String takeTimeout(long timeout) {
+    return retrieveWithTimeout(this::take, timeout);
+  }
+
+  @Override
+  public synchronized String takeTimeout(Predicate<String> filter, long timeout) {
+    return retrieveWithTimeout(() -> take(filter), timeout);
+  }
+
+  @Override
+  public synchronized void clear() {
+    queue.clear();
+  }
+
+  private synchronized String retrieveWithTimeout(Supplier<String> dataSupplier, long timeout) {
+    long start = System.currentTimeMillis();
+    String date = dataSupplier.get();
+
+    try {
+      while (date == null && System.currentTimeMillis() - start < timeout) {
+        wait(timeout);
+        date = dataSupplier.get();
+      }
+    } catch (InterruptedException e) {
+      return null;
+    }
```

Comment:
I'm not sure I fully understand this implementation. it seems that we're always waiting all the way until timeout unless the data was retrieved immediately, is that the desired behavior?

Reply:
In java, wait is different from a sleep.

A wait can be notified by notify() or notifyAll() and it will unlock.
The notifyAll is called when a new message is received and added to the buffer.

I was not sure myself if this was working and thus I wrote the Tests (MessageQueueTest).

Reply:
Thanks. I learned something